sdk/identity/azure-identity/src/managed_identity_source.cpp

```diff
@@ -35,41 +35,46 @@ Azure::Core::Url ManagedIdentitySource::ParseEndpointUrl(
       std::string("The environment variable ") + envVarName + " contains an invalid URL.");
 }
 
+template <typename T>
 std::unique_ptr<ManagedIdentitySource> AppServiceManagedIdentitySource::Create(
     std::string const& clientId,
-    Azure::Core::Credentials::TokenCredentialOptions const& options)
+    Azure::Core::Credentials::TokenCredentialOptions const& options,
+    const char* endpointVarName,
+    const char* secretVarName)
 {
-  constexpr auto EndpointVarName = "MSI_ENDPOINT";
-  auto msiEndpoint = Environment::GetVariable(EndpointVarName);
-  auto msiSecret = Environment::GetVariable("MSI_SECRET");
+  auto msiEndpoint = Environment::GetVariable(endpointVarName);
+  auto msiSecret = Environment::GetVariable(secretVarName);
 
   return (msiEndpoint.empty() || msiSecret.empty())
       ? nullptr
-      : std::unique_ptr<ManagedIdentitySource>(new AppServiceManagedIdentitySource(
-          clientId, options, ParseEndpointUrl(msiEndpoint, EndpointVarName), msiSecret));
+      : std::unique_ptr<ManagedIdentitySource>(
+          new T(clientId, options, ParseEndpointUrl(msiEndpoint, endpointVarName), msiSecret));
```

Comment:
I got nervous whenever I read `new` 🐺   .  Consider using `std::make_unique` instead  ( `std::make_unique<T>(arg,arg2,)`)

Reply:
1. `_detail::ManagedIdentitySource` is not visible to customer. It only has a forward declaration. So the factory methods are not part of public API.

2. `std::make_unique` does require constructor to be public, but so does `new T`. That's why there's "`friend AppServiceManagedIdentitySource`" declaration in `AppServiceV2019ManagedIdentitySource` and in `V2017`. Again, all strictly internal, not visible anywhere, so `friend` is even less of a problem.

Reply:
I'll see if I can use `make_unique`.

Reply:
Ah yes - then it is the `make_unique` that I'll need to declare a friend. I'd avoid that, that declaration would be cumbersome.